src/node_api.cc

```diff
@@ -2461,3 +2463,139 @@ napi_status napi_get_typedarray_info(napi_env env,
 
   return GET_RETURN_STATUS();
 }
+
+namespace uvimpl {
+
+napi_status ConvertUVErrorCode(int code) {
+  switch (code) {
+  case 0:
+    return napi_ok;
+  case UV_EINVAL:
+    return napi_invalid_arg;
+  case UV_ECANCELED:
+    return napi_cancelled;
+  }
+
+  return napi_generic_failure;
+}
+
+// Wrapper around uv_work_t which calls user-provided callbacks.
+class Work {
+ private:
+  explicit Work(napi_env env,
+                napi_async_execute_callback execute = nullptr,
+                napi_async_complete_callback complete = nullptr,
+                void* data = nullptr)
+    : _env(env),
+    _data(data),
+    _execute(execute),
+    _complete(complete) {
+    _request = new uv_work_t;
+    _request->data = this;
+  }
+
+  ~Work() {
+    delete _request;
+  };
+
+ public:
+  static Work* New(napi_env env,
+                   napi_async_execute_callback execute,
+                   napi_async_complete_callback complete,
+                   void* data) {
+    return new Work(env, execute, complete, data);
+  }
+
+  static void Delete(Work* work) {
+    delete work;
+  }
+
+  static void ExecuteCallback(uv_work_t* req) {
+    Work* work = reinterpret_cast<Work*>(req->data);
```

Comment:
static_cast?

Also, if you turn `_request` into a plain data member, you could forget about `req->data` and use Node’s evil `ContainerOf` magic, which is nice because it avoids an unnecessary pointer lookup. (Not that performance really matters here or anything :smile:)